src/OpenLoco/src/Ui/Chart.cpp

```diff
@@ -0,0 +1,297 @@
+
+        auto dataIndex = 0U;
+        std::byte* dataPtr = gs.yData[lineIndex];
+        if ((gs.flags & GraphFlags::dataFrontToBack) != GraphFlags::none)
+        {
+            dataIndex = gs.dataStart[lineIndex];
+            dataPtr = &dataPtr[gs.dataTypeSize * (gs.dataEnd - dataIndex)];
+        }
+
+        while (dataIndex < gs.dataEnd)
+        {
+            if ((gs.flags & GraphFlags::dataFrontToBack) != GraphFlags::none)
+            {
+                dataPtr -= gs.dataTypeSize;
+            }
+
+            int64_t value = graphGetValueFromPointer(dataPtr, gs.dataTypeSize);
+            value >>= gs.numValueShifts;
+
+            int16_t xPos = gs.canvasLeft + dataIndex * gs.xAxisTickIncrement;
+            int16_t yPos = gs.height - value - gs.yOffset;
+
+            if ((gs.flags & GraphFlags::showNegativeValues) != GraphFlags::none)
+            {
+                yPos -= gs.canvasHeight / 2;
+            }
+
+            yPos += gs.top;
+
+            if (_dword_113658C != 1)
+            {
+                auto colour = gs.lineColour[lineIndex];
+                drawingCtx.drawRect(xPos, yPos, 1, 1, colour, Gfx::RectFlags::none);
+
+                auto targetPos = Ui::Point(xPos, yPos);
+                if (previousPos.x != -1)
+                {
+                    drawingCtx.drawLine(previousPos, targetPos, colour);
+                }
+                previousPos = targetPos;
+            }
+            else
+            {
+                auto colour = gs.lineColour[lineIndex];
+                drawingCtx.drawRect(xPos, yPos, 2, 2, colour, Gfx::RectFlags::none);
+            }
+
+            dataIndex++;
+
+            if ((gs.flags & GraphFlags::dataFrontToBack) == GraphFlags::none)
+            {
+                dataPtr += gs.dataTypeSize;
+            }
+        }
+    }
+
+    // 0x004CF824
+    void drawGraph(GraphSettings& gs, Window* self, Gfx::DrawingContext& drawingCtx)
+    {
+        gs.canvasLeft = gs.xOffset + gs.left;
+        gs.canvasHeight = gs.height - gs.yOffset;
+
+        // TODO: unused? remove?
+        gs.canvasBottom = gs.top + gs.height - gs.yOffset;
+
+        int64_t maxValue = graphGetMaxValue(gs);
+
+        // 0x004CFA02
+
+        auto height = gs.canvasHeight;
+        if ((gs.flags & GraphFlags::showNegativeValues) != GraphFlags::none)
+        {
+            height >>= 1;
+        }
+
+        // We work out the number of shifts required to bring a 64bit value
+        // into something that is within the height range
+        gs.numValueShifts = 0;
+        for (auto adjustedMaxValue = maxValue; adjustedMaxValue > height; adjustedMaxValue >>= 1)
+        {
+            gs.numValueShifts++;
+        }
+
+        if ((gs.flags & GraphFlags::hideAxesAndLabels) == GraphFlags::none)
+        {
+            graphDrawAxesAndLabels(gs, self, drawingCtx);
+        }
+
+        // 0x004CFD59
+        for (_dword_113658C = 0; _dword_113658C < 2; _dword_113658C++) // iteration/pass??
```

Comment:
What would be a good name for this?

I think it's a bit strange how it alternates between drawing points and lines, tbh. Initially, I thought it would make the points more visible for lines with overlap, but that doesn't appear to be the case.

Reply:
Actually, it seems `byte_113DD99` is only ever set to 1, so the points (flag 0) are never even drawn...